main.vala

```diff
@@ -639,291 +514,1289 @@ class Vls.Server {
                 sb.insert ((ssize_t) pos, changeEvent.text);
             }
         }
-        source.file.content = sb.str;
+        source.content = sb.str;
 
-        // if we're at this point, the file is present in the context
-        // any change we make invalidates the context
-        ctx.invalidate ();
+        source.compilation.invalidate ();
+        request_context_update (client);
+    }
 
-        // we have to update everything
-        ctx.check ();
+    void request_context_update (Jsonrpc.Client client) {
+        update_context_client = client;
+        update_context_requests += 1;
+        int64 delay_us = int64.min (update_context_delay_inc_us * update_context_requests, update_context_delay_max_us);
+        update_context_time_us = get_monotonic_time () + delay_us;
+        debug (@"Context(s) update (re-)scheduled in $((int) (delay_us / 1000)) ms");
+    }
 
-        publishDiagnostics (client);
+    void check_update_context () {
+        if (update_context_requests > 0 && get_monotonic_time () >= update_context_time_us) {
+            update_context_requests = 0;
+            update_context_time_us = 0;
+            /* This must come after the resetting of the two variables above,
+             * since it's possible for publishDiagnostics to eventually call
+             * one of our JSON-RPC callbacks through g_main_context_iteration (),
+             * if we get a new message while sending the textDocument/publishDiagnostics
+             * notifications. */
+            foreach (var target in builds) {
+                target.compile ();
+                publishDiagnostics (target, update_context_client);
+            }
+        }
     }
 
-    void publishDiagnostics (Jsonrpc.Client client, string? doc_uri = null) {
-        Collection<TextDocument> docs;
-        TextDocument? doc = doc_uri == null ? null : ctx.get_source_file (doc_uri);
+    delegate void OnContextUpdatedFunc (bool request_cancelled);
+
+    /**
+     * Rather than satisfying all requests in `check_update_context ()`,
+     * to avoid race conditions, we have to spawn a timeout to check for 
```

Comment:
What race conditions? Can you have race conditions in a single thread?

Reply:
Actually you can (an example would be a signal handler that modifies critical data) but I wrote this before taking a look at `GLib.MainLoop`, so this comment is wrong.

Reply:
can you update the comment then?